client.html

```diff
@@ -53,6 +78,221 @@
+                if (branchIndicatorsElems[i] != ev.target) {
+                    branchOriginsElems.push(branchIndicatorsElems[i]);
+                }
+            }
+
+            // use a different origin as main scroll-to element each click
+            var scrollMainElemIdx;
+            if ('nextScrollMainElemIdx' in ev.target.dataset) {
+                scrollMainElemIdx = parseInt(ev.target.dataset.nextScrollMainElemIdx);
+            } else {
+                scrollMainElemIdx = 0;
+            }
+            scrollMainElemIdx %= branchOriginsElems.length;
+            ev.target.dataset.nextScrollMainElemIdx = scrollMainElemIdx + 1;
+
+            var scrollMainElem = branchOriginsElems[scrollMainElemIdx];
+            var scrollOtherElems = [ev.target];
+            for (var i = 0; i < branchOriginsElems.length; i++) {
+                if (i != scrollMainElemIdx) {
+                    scrollOtherElems.push(branchOriginsElems[i]);
+                }
+            }
+            scrollIntoView(scrollMainElem, scrollOtherElems);
+
+            highlightBranchClearAll();
+            highlightBranchStay(ev.target.dataset.branchesClass);
+        }
+
+        function onMouseEnterBranchIndicator(ev) {
+            highlightBranchTemp(ev.target.dataset.branchesClass,);
+        }
+
+        function onMouseLeaveBranchIndicator(ev) {
+            highlightBranchTempClear(ev.target.dataset.branchesClass);
+        }
+
+        function setDiffHtml(diffHtml) {
+            ASMDW.dom.diffContainer.innerHTML = diffHtml;
+            var tbody = ASMDW.dom.diffContainer.getElementsByTagName('tbody')[0];
+            for (var i = 0; i < tbody.children.length; i++) {
+                var tr = tbody.children[i];
+                for (var j = 0; j < 2 && j < tr.children.length; j++) {
+                    var td = tr.children[j];
+                    var spanChildren = td.getElementsByTagName('span');
+                    var wrapBranchIndicatorsElems = [];
+                    for (var k = 0; k < spanChildren.length; k++) {
+                        var spanChild = spanChildren[k];
+                        if ('branchesClass' in spanChild.dataset) {
+                            // branch origins have data-branch-target
+                            if ('branchTarget' in spanChild.dataset) {
+                                spanChild.addEventListener('click', onClickBranchOrigin);
+                            } else {
+                                spanChild.addEventListener('click', onClickBranchTarget);
+                            }
+                            spanChild.addEventListener('mouseenter', onMouseEnterBranchIndicator);
+                            spanChild.addEventListener('mouseleave', onMouseLeaveBranchIndicator);
+                            spanChild.addEventListener('animationend',
+                                ev => ev.target.classList.remove('highlight-branch'));
```

Comment:
instead of setting event listeners on individual elements like this, I'd suggest setting a single one on document and having it check what to do depending on event.target

Reply:
done for `click`, but `mouseenter`/`mouseleave` need to be on individuals elements from what I understood

Reply:
It works if you use a capturing event listener (addEventListener(..., true)). May be that performance suffers